toonz/sources/toonz/xsheetdragtool.cpp

```diff
@@ -1595,22 +1597,55 @@ class ColumnMoveDragTool final : public XsheetGUI::DragTool {
     if (indices.empty()) return;
     m_firstCol = m_lastCol = *indices.begin();
     assert(m_firstCol >= 0);
-    m_offset = m_firstCol - col;
+    m_origOffset = m_offset = m_firstCol - col;
     assert(m_lastCol == *indices.begin());
     getViewer()->update();
   }
   void onDrag(const CellPosition &pos) override {
     int col                     = pos.layer();
     TColumnSelection *selection = getViewer()->getColumnSelection();
-    std::set<int> indices       = selection->getIndices();
+    TApp *app                   = TApp::instance();
+    TXsheet *xsh                = app->getCurrentXsheet()->getXsheet();
+
+    std::set<int> indices = selection->getIndices();
     if (indices.empty()) return;
 
     assert(m_lastCol == *indices.begin());
 
+    int origCol      = col;
     if (col < 0) col = 0;
-    if (col == m_lastCol) return;
-    int dCol  = col - m_lastCol;
-    m_lastCol = col;
+    int dCol         = col - (m_lastCol - m_offset);
+
+    if (!getViewer()->orientation()->isVerticalTimeline()) {
+      if (origCol < 0) dCol += origCol;
+      std::set<int> ii;
+      int newBegin = *indices.begin() + dCol;
+      if (newBegin < 0) {
+        newBegin *= -1;
+        for (int x = 0; x < newBegin; x++) ii.insert(x);
+        ColumnCmd::insertEmptyColumns(ii);
```

Comment:
IMO the undo for inserting empty columns while column dragging operation should be put in a block.
How about adding the lines as follows?:
```
@@ -1600,6 +1600,9 @@ public:
     m_origOffset = m_offset = m_firstCol - col;
     assert(m_lastCol == *indices.begin());
     getViewer()->update();
+
+    if (!getViewer()->orientation()->isVerticalTimeline())
+      TUndoManager::manager()->beginBlock();
   }
   void onDrag(const CellPosition &pos) override {
     int col                     = pos.layer();

@@ -1658,14 +1666,17 @@ public:
   }
   void onRelease(const CellPosition &pos) override {
     int delta = m_lastCol - m_firstCol;
-    if (delta == 0) return;
-    TColumnSelection *selection = getViewer()->getColumnSelection();
-    std::set<int> indices       = selection->getIndices();
-    if (!indices.empty()) {
-      TUndoManager::manager()->add(new ColumnMoveUndo(indices, delta));
-      TApp::instance()->getCurrentScene()->setDirtyFlag(true);
-      TApp::instance()->getCurrentXsheet()->notifyXsheetChanged();
+    if (delta != 0) {
+      TColumnSelection *selection = getViewer()->getColumnSelection();
+      std::set<int> indices       = selection->getIndices();
+      if (!indices.empty()) {
+        TUndoManager::manager()->add(new ColumnMoveUndo(indices, delta));
+        TApp::instance()->getCurrentScene()->setDirtyFlag(true);
+        TApp::instance()->getCurrentXsheet()->notifyXsheetChanged();
+      }
     }
+    if (!getViewer()->orientation()->isVerticalTimeline())
+      TUndoManager::manager()->endBlock();
   }
 };
```

Reply:
Added. Great suggestion. Thanks